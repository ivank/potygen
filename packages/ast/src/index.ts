import { Parser } from '@ikerin/rd-parse';
import { Grammar } from './grammar';
import { QueryTag, WithTag } from './grammar.types';

export {
  AnyCastTag,
  AnyTypeTag,
  ArrayConstructorTag,
  ArrayIndexRangeTag,
  ArrayIndexTag,
  AsTag,
  BetweenTag,
  BinaryExpressionTag,
  BinaryOperatorTag,
  BooleanTag,
  CaseTag,
  CastableDataTypeTag,
  CastTag,
  CollateTag,
  ColumnsTag,
  ColumnTag,
  CombinationTag,
  ComparationExpressionTag,
  ComparationOperatorTag,
  ConflictConstraintTag,
  ConflictTag,
  ConflictTargetTag,
  ConstantTag,
  CountTag,
  CTETag,
  DataTypeTag,
  DefaultTag,
  DeleteTag,
  DistinctTag,
  DoNothingTag,
  DoUpdateTag,
  ElseTag,
  ExpressionListTag,
  ExpressionTag,
  FromListItemTag,
  FromListTag,
  FromTag,
  FunctionTag,
  GroupByTag,
  HavingTag,
  IdentifierTag,
  InsertTag,
  JoinOnTag,
  JoinTag,
  JoinTypeTag,
  JoinUsingTag,
  LimitTag,
  NullTag,
  NumberTag,
  OffsetTag,
  OperatorExpressionTag,
  OrderByItemTag,
  OrderByTag,
  OrderDirectionTag,
  ParameterTag,
  PgCastTag,
  QuotedNameTag,
  ReturningListItemTag,
  ReturningTag,
  RowTag,
  SelectListItemTag,
  SelectListTag,
  SelectParts,
  SelectTag,
  SetItemTag,
  SetListTag,
  SetMapTag,
  SetTag,
  SqlTag,
  StarIdentifierTag,
  StarTag,
  StringTag,
  TableTag,
  Tag,
  TypeArrayTag,
  TypeTag,
  UnaryExpressionTag,
  UnaryOperatorTag,
  UpdateFromTag,
  UpdateTag,
  UsingTag,
  ValuesListTag,
  ValuesTag,
  WhenTag,
  WhereTag,
  WithTag,
} from './grammar.types';

export {
  isAnyCast,
  isAnyType,
  isArrayConstructor,
  isArrayIndex,
  isArrayIndexRange,
  isAs,
  isBetween,
  isBinaryExpression,
  isBinaryOperator,
  isBoolean,
  isCase,
  isCast,
  isCastableDataType,
  isCollateTag,
  isColumn,
  isColumns,
  isCombination,
  isComparationExpression,
  isComparationOperator,
  isConflict,
  isConflictConstraint,
  isConflictTarget,
  isConstant,
  isCount,
  isCTETag,
  isDataType,
  isDefault,
  isDelete,
  isDistinct,
  isDoNothing,
  isDoUpdate,
  isElse,
  isExpression,
  isExpressionList,
  isFrom,
  isFromList,
  isFromListItem,
  isFunction,
  isGroupBy,
  isHaving,
  isIdentifier,
  isInsertTag,
  isJoin,
  isJoinOn,
  isJoinType,
  isJoinUsing,
  isLimit,
  isNull,
  isNumber,
  isOffset,
  isOperatorExpression,
  isOrderBy,
  isOrderByItem,
  isOrderDirection,
  isParameter,
  isPgCast,
  isQuotedName,
  isReturning,
  isReturningListItem,
  isRow,
  isSelect,
  isSelectList,
  isSelectListItem,
  isSelectTag,
  isSet,
  isSetItem,
  isSetList,
  isSetMap,
  isStar,
  isStarIdentifier,
  isString,
  isTable,
  isType,
  isTypeArray,
  isUnaryExpression,
  isUnaryOperator,
  isUpdate,
  isUpdateFrom,
  isUsing,
  isValues,
  isValuesList,
  isWhen,
  isWhere,
  isWithTag,
  isWrappedExpression,
} from './grammar.guards';

export {
  chunk,
  first,
  groupBy,
  identity,
  initial,
  isDiffBy,
  isEmpty,
  isEqual,
  isNil,
  isObject,
  isUnique,
  isUniqueBy,
  last,
  length,
  orderBy,
  tail,
  toMilliseconds,
} from './util';

export { Grammar } from './grammar';

export const parser = Parser<QueryTag | WithTag>(Grammar);
