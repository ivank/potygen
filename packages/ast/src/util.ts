export const length = (item: unknown): number | undefined => (Array.isArray(item) ? item.length : undefined);

export const identity = <T>(item: T): T => item;

export const isEmpty = <T>(items?: T[]): boolean => !items || items?.length === 0;

export function tail<T>(items: [unknown, ...T[]]): T[];
export function tail<T>(items: T[]): T[];
export function tail<T>(items: T[]): T[] {
  return items.slice(1);
}

export function last<T>(items?: [...unknown[], T]): T;
export function last<T>(items?: T[]): T | undefined;
export function last<T>(items?: T[]): T | undefined {
  return items?.[items.length - 1];
}

export function first<T>(items?: [T, ...unknown[]]): T;
export function first<T>(items?: T[]): T | undefined;
export function first<T>(items?: T[]): T | undefined {
  return items?.[0];
}

export function initial<T>(items: [...T[], unknown]): T[];
export function initial<T>(items: T[]): T[];
export function initial<T>(items: T[]): T[] {
  return items.slice(0, -1);
}

export const isUnique =
  <T>(predicate: (item: T) => unknown = identity) =>
  (item: T, index: number, items: T[]): boolean =>
    items.findIndex((current) => predicate(item) === predicate(current)) === index;

export const isUniqueBy =
  <T>(predicate: (a: T, b: T) => boolean = isEqual) =>
  (item: T, index: number, items: T[]): boolean =>
    items.findIndex((current) => predicate(item, current)) === index;

export const orderBy =
  <T>(predicate: (item: T) => any = identity) =>
  (a: T, b: T): number => {
    const orderA = predicate(a);
    const orderB = predicate(b);
    return orderA === orderB ? 0 : orderA > orderB ? 1 : -1;
  };

export const isNil = <T>(item: T | undefined | null): item is T => Boolean(item);

export const isDiffBy =
  <T>(predicate: (item: T) => unknown, to: T[]) =>
  (fromItem: T): boolean =>
    !to.some((toItem) => predicate(fromItem) === predicate(toItem));

export const groupBy = <T, K extends string | number>(predicate: (item: T) => K, items: T[]): Record<K, T[]> =>
  items.reduce((acc, item) => {
    const key = predicate(item);
    return { ...acc, [key]: (acc[key] ?? ([] as T[])).concat(item) };
  }, {} as Record<K, T[]>);

export const chunk = <T>(size: number, items: T[]): T[][] =>
  [...Array(Math.ceil(items.length / size))].map((_, index) => items.slice(index * size, (index + 1) * size));

export const isObject = (item: unknown): item is Record<string, unknown> => typeof item === 'object' && item !== null;

export const isEqual = (a: unknown, b: unknown): boolean => {
  if (a === b) {
    return true;
  }
  if (a === undefined || b === undefined) {
    return false;
  }
  if (typeof a === 'object' && typeof b === 'object') {
    if (a === null || b === null) {
      return false;
    } else if (a instanceof Date && b instanceof Date) {
      return a.getTime() === b.getTime();
    } else if (Array.isArray(a) && Array.isArray(b)) {
      return a.length === b.length && a.every((item, index) => isEqual(item, b[index]));
    } else if (isObject(a) && isObject(b)) {
      const aKeys = Object.keys(a).sort();
      const bKeys = Object.keys(a).sort();
      return isEqual(aKeys, bKeys) && aKeys.every((key) => isEqual(a[key], b[key]));
    }
  }
  return false;
};
