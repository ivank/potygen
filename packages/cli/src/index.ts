export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
