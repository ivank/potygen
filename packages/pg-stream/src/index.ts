export { toInternalCursorGenerator } from './internal-cursor-generator';
export { ToEachBatch, ToEachBatchOptions, toEachBatch } from './each-batch';
export { ToReadable, ToReadableOptions, toReadable } from './readable';
export {
  ToAsyncIterator,
  toAsyncIterator,
  ToIteratorOptions,
  ToAsyncBatchIterator,
  toAsyncBatchIterator,
} from './iterator';
