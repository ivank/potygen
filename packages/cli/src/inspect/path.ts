import { Tag, isNode, SqlTag } from '@potygen/ast';
import { PathItem, Path } from '../types';

const isItem =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (item: PathItem<Tag>): item is PathItem<T> =>
    predicate(item.tag);

export const closestParent =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): T | undefined =>
    path.find(isItem(predicate))?.tag;

export const closestParentPath =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): PathItem<T> | undefined =>
    path.find(isItem(predicate));

export const toPath = (tag: Tag, offset: number, path: Path = []): Path | undefined => {
  if (isNode(tag)) {
    const index = (tag.values as Tag[]).findIndex((item) => offset >= item.start && offset <= item.end + 1);

    if (index !== -1) {
      return toPath(tag.values[index], offset, [{ tag, index }, ...path]);
    } else {
      return [{ tag }, ...path];
    }
  }
  return [{ tag }, ...path];
};
