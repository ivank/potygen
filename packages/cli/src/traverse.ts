import { readFileSync, watchFile } from 'fs';
import { Readable, Writable } from 'stream';
import { glob } from './glob';
import {
  createSourceFile,
  ScriptTarget,
  isTaggedTemplateExpression,
  Node,
  isImportDeclaration,
  isStringLiteral,
  isNoSubstitutionTemplateLiteral,
  isNamedImports,
  SourceFile,
  isIdentifier,
  isCallExpression,
} from 'typescript';
import { basename, relative } from 'path';
import {
  parser,
  QueryInterface,
  toQueryInterface,
  loadQueryInterfacesData,
  toLoadedQueryInterface,
  LoadError,
  ParsedSqlFileLoadError,
  ParsedTypescriptFileLoadError,
  ParseError,
  LoadContext,
  LoadedData,
  LoadedFile,
  Logger,
  ParsedFile,
  ParsedSqlFile,
  ParsedTypescriptFile,
  TemplateTagQuery,
  loadAllData,
} from '@potygen/potygen';
import { ClientBase } from 'pg';
import { emitLoadedFile } from './emit';
import { inspect } from 'util';

const toTemplateParent = (node: Node): Node =>
  isCallExpression(node.parent) ? toTemplateParent(node.parent) : node.parent;

const getTemplateTagQueries = (ast: SourceFile): TemplateTagQuery[] => {
  const queries: TemplateTagQuery[] = [];
  let tagPropertyName = 'sql';

  const visitor = (node: Node): void => {
    if (
      isImportDeclaration(node) &&
      isStringLiteral(node.moduleSpecifier) &&
      node.importClause?.namedBindings &&
      isNamedImports(node.importClause.namedBindings) &&
      node.moduleSpecifier.text === '@potygen/potygen'
    ) {
      tagPropertyName =
        node.importClause?.namedBindings.elements.find(
          ({ propertyName, name }) => (propertyName ?? name).text === 'sql',
        )?.name.text ?? tagPropertyName;
    } else if (
      isTaggedTemplateExpression(node) &&
      isNoSubstitutionTemplateLiteral(node.template) &&
      isIdentifier(node.tag) &&
      node.tag.text === tagPropertyName
    ) {
      const tag = {
        name: toTemplateParent(node).getChildAt(0).getText(),
        pos: node.template.pos + 1,
        template: node.template.text,
      };
      try {
        queries.push({ ...tag, queryInterface: toQueryInterface(parser(node.template.text).ast) });
      } catch (error) {
        throw new ParseError(tag, `Error parsing sql: ${error instanceof Error ? error.message : String(error)}`);
      }
    } else {
      node.forEachChild(visitor);
    }
  };

  visitor(ast);
  return queries;
};

const toParsedTypescriptFile = (path: string): ParsedTypescriptFile => {
  const sourceText = readFileSync(path, 'utf-8');
  const source = createSourceFile(basename(path), sourceText, ScriptTarget.ES2021, true);
  return { type: 'ts', source, path, queries: getTemplateTagQueries(source) };
};

const toParsedSqlFile = (path: string): ParsedSqlFile | undefined => {
  const content = readFileSync(path, 'utf-8');
  return { type: 'sql', path, content, queryInterface: toQueryInterface(parser(content).ast) };
};

const toQueryInterfaces = (files: ParsedFile[]): QueryInterface[] =>
  files.flatMap((file) =>
    file.type === 'ts' ? file.queries.map((query) => query.queryInterface) : file.queryInterface,
  );

const loadDataFromParsedFiles = async (
  ctx: LoadContext,
  data: LoadedData[],
  files: ParsedFile[],
): Promise<LoadedData[]> => loadQueryInterfacesData(ctx, toQueryInterfaces(files), data);

const isError = (error: unknown): error is LoadError | ParseError =>
  error instanceof LoadError || error instanceof ParseError;

const loadFile =
  (data: LoadedData[]) =>
  (file: ParsedFile): LoadedFile => {
    if (file.type === 'sql') {
      try {
        return { ...file, loadedQuery: toLoadedQueryInterface(data)(file.queryInterface) };
      } catch (error) {
        throw isError(error) ? new ParsedSqlFileLoadError(file, error) : error;
      }
    } else {
      return {
        ...file,
        queries: file.queries.map((template) => {
          try {
            return { ...template, loadedQuery: toLoadedQueryInterface(data)(template.queryInterface) };
          } catch (error) {
            throw isError(error) ? new ParsedTypescriptFileLoadError(file, template, error) : error;
          }
        }),
      };
    }
  };

const parseFile = (path: string): ParsedFile | undefined => {
  if (path.endsWith('.ts')) {
    const file = toParsedTypescriptFile(path);
    return file.queries.length > 0 ? file : undefined;
  } else {
    return toParsedSqlFile(path);
  }
};

export class SqlRead extends Readable {
  public source: Generator<string, void, unknown>;
  public watchedFiles = new Set<string>();

  constructor(public options: { path: string; root: string; watch: boolean; logger: Logger }) {
    super({ objectMode: true });
    this.source = glob(options.path, options.root);
  }

  next() {
    let path: IteratorResult<string>;
    while (!(path = this.source.next()).done) {
      const file = parseFile(path.value);
      if (file) {
        return file;
      }
    }
    return undefined;
  }

  watchFile(path: string): () => void {
    return () => {
      const file = parseFile(path);
      if (file) {
        this.options.logger.info(`Processing ${relative(this.options.root ?? '.', path)}`);
        this.push(file);
      }
    };
  }

  _read() {
    const next = this.next();
    if (next) {
      if (this.options.watch && !this.watchedFiles.has(next.path)) {
        this.watchedFiles.add(next.path);
        watchFile(next.path, this.watchFile(next.path));
      }
      this.options.logger.info(`Processing ${relative(this.options.root ?? '.', next.path)}`);
      this.push(next);
    } else if (!this.options.watch) {
      this.options.logger.info(`Done`);
      this.push(null);
    }
  }
}

export class QueryLoader extends Writable {
  public ctx: LoadContext;
  public data: LoadedData[] = [];
  constructor(
    public options: {
      db: ClientBase;
      root: string;
      template: string;
      logger: Logger;
      typePrefix?: string;
      preload?: boolean;
    },
  ) {
    super({ objectMode: true });
    this.ctx = { db: options.db, logger: options.logger ?? console };
  }

  async _writev(
    chunks: Array<{ chunk: ParsedFile; encoding: BufferEncoding }>,
    callback: (error?: Error | null) => void,
  ): Promise<void> {
    try {
      const parsedFiles = chunks.map((file) => file.chunk);
      this.ctx.logger.debug(
        `Parse files: ${inspect(
          parsedFiles.map((file) => `${relative(this.options.root, file.path)} (${file.type})`),
        )}`,
      );
      this.data = this.options.preload
        ? this.data.length === 0
          ? await loadAllData(this.ctx, this.data)
          : this.data
        : await loadDataFromParsedFiles(this.ctx, this.data, parsedFiles);
      await Promise.all(
        parsedFiles
          .map(loadFile(this.data))
          .map(emitLoadedFile(this.options.root, this.options.template, this.options.typePrefix)),
      );
    } catch (error) {
      this.options.logger.error(error instanceof Error ? String(error) : new Error(String(error)));
    }
    callback();
  }

  async _write(
    file: ParsedTypescriptFile,
    encoding: BufferEncoding,
    callback: (error?: Error | null) => void,
  ): Promise<void> {
    try {
      this.ctx.logger.debug(`Parse file: ${relative(this.options.root, file.path)} (${file.type})`);
      this.data = this.options.preload
        ? this.data.length === 0
          ? await loadAllData(this.ctx, this.data)
          : this.data
        : await loadDataFromParsedFiles(this.ctx, this.data, [file]);
      await emitLoadedFile(
        this.options.root,
        this.options.template,
        this.options.typePrefix,
      )(loadFile(this.data)(file));
    } catch (error) {
      this.options.logger.error(error instanceof Error ? String(error) : new Error(String(error)));
    }
    callback();
  }
}
