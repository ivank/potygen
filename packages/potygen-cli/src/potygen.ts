import { Command, createCommand } from 'commander';
import { Logger } from './types';
import { QueryLoader, SqlRead } from './traverse';
import { Client } from 'pg';
import { promisify } from 'util';
import { pipeline } from 'stream';
import { existsSync, readFileSync } from 'fs';
import { Record, String, Optional, Boolean, Static } from 'runtypes';

const Config = Record({
  files: Optional(String),
  watch: Optional(Boolean),
  root: Optional(String),
  connection: Optional(String),
  template: Optional(String),
});

type ConfigType = Static<typeof Config>;

const asyncPipeline = promisify(pipeline);

export const potygen = (logger: Logger = console): Command =>
  createCommand('potygen')
    .description('Convert postgres query files inty typescript types')
    .option('-c, --config <config>', 'A configuration file to load', 'potygen.config.json')
    .option('-f, --files <files>', 'A glob pattern to search files by (default: "**/*.sql")')
    .option('-w, --watch', 'Watch for file changes and update live')
    .option('-r, --root <root>', `Set the root directory (default: ${process.cwd()})`)
    .option(
      '-n, --connection <connection>',
      'Connection to the postgres database. URI (default: "postgres://localhost:5432/db")',
    )
    .option(
      '-t, --template <template>',
      'A template of the path, where to generate the typescript type files. The parameters are the response from node\'s path.parse function (default: "{{dir}}/{{name}}.queries.ts")',
    )
    .action(async (options: ConfigType & { config: string }) => {
      const config: ConfigType = existsSync(options.config)
        ? Config.check(JSON.parse(readFileSync(options.config, 'utf-8')))
        : {};

      const { root, connection, watch, files, template } = {
        files: '**/*.sql',
        root: process.cwd(),
        template: '{{dir}}/{{name}}.queries.ts',
        connection: 'postgres://localhost:5432/db',
        watch: false,
        ...config,
        ...options,
      };

      if (process.env.POTYGEN_DEBUG) {
        console.info('Potygen Config', { root, connection, watch, files, template });
      }

      const db = new Client(connection);
      await db.connect();
      try {
        const sqls = new SqlRead({ path: files, root, watch, logger });
        const sink = new QueryLoader({ db, root, template });

        logger.info(`Potygen started processing ("${files}", watch: ${watch})`);

        await asyncPipeline(sqls, sink);
      } catch (error) {
        if (process.env.POTYGEN_DEBUG && error instanceof Error) {
          logger.error(error.stack);
        } else {
          logger.error(global.String(error));
        }
      } finally {
        await db.end();
      }
    });
