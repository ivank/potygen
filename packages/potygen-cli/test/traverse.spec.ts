import { readFileSync } from 'fs';
import { join, relative } from 'path';
import { Client } from 'pg';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { glob } from '../src';
import { SqlRead, QueryLoader } from '../src/traverse';

const asyncPipeline = promisify(pipeline);

describe('Traverse', () => {
  it('Should work', async () => {
    const db = new Client({ database: 'potygen', user: 'potygen', password: 'dev-pass' });
    try {
      await db.connect();

      const logger = { info: jest.fn(), error: jest.fn(), debug: jest.fn() };
      const sqls = new SqlRead({
        path: '*.sql',
        root: join(__dirname, '../../../sql'),
        logger,
        watch: false,
      });
      const sink = new QueryLoader({
        db,
        root: __dirname,
        template: join(__dirname, '__generated__/{{name}}.queries.ts'),
        logger,
      });

      await asyncPipeline(sqls, sink);

      for (const file of Array.from(glob('__generated__/*.queries.ts', __dirname))) {
        expect(readFileSync(file, 'utf-8')).toMatchSnapshot(relative(__dirname, file));
      }
    } finally {
      await db.end();
    }
  });
});
