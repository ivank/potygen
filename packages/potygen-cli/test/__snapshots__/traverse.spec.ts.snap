// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traverse Should work: __generated__/account-levelisation.update.query.queries.ts 1`] = `
"export interface Params {
    isAccepted: boolean;
    id: number;
}
export interface Result {
    id: number;
    isAccepted: boolean;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/account-levelisation.upsert.query.queries.ts 1`] = `
"export interface Params {
    levelisationId: number;
    start?: Date;
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/account-levelisations.query.queries.ts 1`] = `
"import { Json } from \\"@potygen/query\\";
export interface Params {
    q: string;
    resolvedPostlev: \\"\\" | boolean;
    state: (\\"Pending\\" | \\"Done\\");
    levelisationId: number;
    ids: number[];
    sortField: \\"totalPayment\\" | \\"vatPayment\\" | \\"generationPayment\\" | \\"exportPayment\\" | \\"generationEnergy\\" | \\"exportEnergy\\" | \\"cfrFitId\\" | \\"state\\" | \\"isAccepted\\" | \\"error\\" | \\"exportType\\";
    sortOrder: \\"DESC\\" | \\"ASC\\";
    limit: number;
    offset: number;
}
export interface Result<TGenerationPeriods = unknown, TExportPeriods = unknown, TErrorParams = unknown> {
    id: number;
    levelisationId: number;
    accountId: number;
    installationId: number;
    state: \\"Pending\\" | \\"Done\\";
    isAccepted: boolean;
    generationStartReadOn?: Date;
    generationStartReadValue?: string;
    generationEndReadOn?: Date;
    generationEndReadValue?: string;
    generationPercentageSplit?: string;
    generationPayment?: string;
    generationEnergy?: string;
    exportStartReadOn?: Date;
    exportStartReadValue?: string;
    exportEndReadOn?: Date;
    exportEndReadValue?: string;
    exportPercentageSplit?: string;
    exportPayment?: string;
    exportEnergy?: string;
    exportType?: \\"Deemed\\" | \\"Metered Export\\" | \\"Off Grid\\" | \\"PPA\\";
    technologyType?: \\"PV\\" | \\"H\\" | \\"W\\" | \\"AD\\" | \\"CHP\\";
    totalPayment?: string;
    vatPayment?: string;
    generationPeriods?: Json<TGenerationPeriods>;
    exportPeriods?: Json<TExportPeriods>;
    errorCode?: string;
    errorParams?: Json<TErrorParams>;
    cfrFitId: string;
    resolvedPostlevId?: number;
    isBacsPaymentsSent?: boolean;
    isChequePaymentsSent?: boolean;
}
export interface Query<TGenerationPeriods = unknown, TExportPeriods = unknown, TErrorParams = unknown> {
    params: Params;
    result: Result<TGenerationPeriods, TExportPeriods, TErrorParams>;
}
"
`;

exports[`Traverse Should work: __generated__/account-levelisations-bulk-close-accounts.query.queries.ts 1`] = `
"export interface Params {
    ids: number[];
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/account-levelisations-set-bulk-update.query.queries.ts 1`] = `
"export interface Params {
    ids: number[];
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/accounts.list.query.queries.ts 1`] = `
"export interface Params {
    q: string | string;
    ids: number[];
}
export interface Result {
    total: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/accounts.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/accounts.query.queries.ts 1`] = `
"export interface Params {
    q: string | string;
    ids: number[];
    limit: number;
    offset: number;
}
export interface Result {
    id: number;
    state: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
    startOn?: Date;
    firstName: string;
    lastName: string;
    email?: string;
    address: string;
    fitIds: string[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/bacs.query.queries.ts 1`] = `
"export interface Params {
    quarter?: string;
    perPage: string;
    offset: string;
}
export interface Result {
    fitId: string;
    name?: string;
    amount: string;
    quarter?: string;
    bankDetails: string;
    addressLine1?: string;
    addressLine2?: string;
    addressLine3?: string;
    postcode: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/contracts.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/delete-accounts.query.queries.ts 1`] = `
"export interface Params {
    type: \\"FIT\\" | \\"SEG\\";
    state: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
}
export interface Result {
    id: number;
    sourceSystemId?: number;
    customerId: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/generation-read-data.query.queries.ts 1`] = `
"export interface Params {
    installationId: number;
    exportDateOn: Date;
}
export interface Result {
    dateOn: Date;
    value: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/installation-meter-update.query.queries.ts 1`] = `
"export interface Params {
    removalDate?: Date;
    meterId: number;
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/installations.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/levelisations.list.query.queries.ts 1`] = `
"export interface Params {
    q: string;
    ids: number[];
    sortField: \\"quarter\\";
    sortOrder: \\"DESC\\" | \\"ASC\\";
    limit: number;
    offset: number;
}
export interface Result {
    id: number;
    quarter: string;
    startOn: Date;
    endOn: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/meter-reads.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/process-batch.query.queries.ts 1`] = `
"export interface Params {
    fitIds: string[];
}
export interface Result {
    cfrFiTId?: string;
    email?: string;
    contactName: string;
    company?: string;
    line1?: string;
    line2: string;
    town?: string;
    postcode: string;
    county?: string;
    country: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/rates-by-tarrif-ids.query.queries.ts 1`] = `
"export interface Params {
    ids: number[][];
}
export interface Result {
    tariffId: number;
    rate: string;
    startOn: Date;
    tariffCode: string;
    tariffType: \\"Export\\" | \\"Generation\\" | \\"SEG\\";
    endOn?: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/read-create.query.queries.ts 1`] = `
"export interface Params {
    meterId: number;
    reason?: string;
    type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    dateOn: Date;
    submittedAt?: Date;
    value: number;
}
export interface Result {
    id: number;
    value: number;
    meterId: number;
    submittedAt?: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/reads-by-account-id.query.queries.ts 1`] = `
"export interface Params {
    id: number;
    offset: string;
    perPage: string;
}
export interface Result {
    id: number;
    msn: string;
    meterId: number;
    dateOn: Date;
    value: number;
    type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    reason?: string;
    createdAt: Date;
    updatedAt?: Date;
    sourceSystemId?: number;
    mpan: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/tables.queries.ts 1`] = `
"export interface Params {
    tableNames: {
        schema: unknown;
        name: unknown;
    }[];
}
export interface Result {
    type: \\"Table\\";
    name: {
        schema: string;
        name: string;
    };
    columns: {
        name: string;
        isNullable: string;
        enum: string;
        type: string;
    }[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/tariff-rates.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;
